keps/sig-storage/596-csi-inline-volumes/README.md

```diff
@@ -213,32 +250,214 @@ Inline volume requests can only participate in mount/unmount volume operations.
 * Enable testing of an external ephemeral CSI driver: https://github.com/kubernetes/kubernetes/pull/79983/files#diff-e5fc8d9911130b421b74b1ebc273f458
 * Enable testing of the csi-host-path-driver in ephemeral mode in Kubernetes-CSI Prow jobs and Kubernetes itself: TODO
 
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind `CSIInlineVolume` feature flag
+- Initial unit tests and e2e tests implemented
+
+#### Beta
+
+- Feature flag enabled by default
+- CSI drivers can support both ephemeral inline volumes and persistent volumes
+
+#### GA
+
+- Remove dependency on deprecated `PodSecurityPolicy` and document new strategy
+- Upgrade / downgrade manual testing
+- Conformance tests implemented
+
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: CSIInlineVolume
+  - Components depending on the feature gate: kubelet
+
+###### Does enabling the feature change any default behavior?
+
+  Enabling the feature gate will **not** change the default behavior.
+  Each CSI driver must opt-in to support ephemeral inline volumes.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+  Yes. Disabling the feature gate will disable support for inline CSI volumes.
+  It can also be disabled on a specific CSI driver by removing the "Ephemeral"
+  VolumeLifecycleMode from `CSIDriver.VolumeLifecycleModes`.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+  Reenabling the feature will again allow CSI drivers that opt-in to be used
+  as ephemeral inline volumes in the pod spec.
+
+###### Are there any tests for feature enablement/disablement?
+
+  Yes, the unit tests will test with the feature gate enabled and disabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+  A rollout should not impact running workloads, since the CSI driver
+  also needs to support ephemeral inline volumes, and the pod spec would need
+  to be updated to make use of the feature.
+
+  A rollback (or disabling the feature flag) may impact running workloads,
+  as any pod specs taking advantage of ephemeral inline volumes would fail
+  to start when the feature is disabled or the CSIDriver does not support it.
+  It is recommended to delete any pods using this feature before disabling it.
+
+###### What specific metrics should inform a rollback?
+
+  No rollback criteria based on metrics alone. This feature has been enabled by
+  default since 1.16 (beta) and pods that are using ephemeral inline volumes should
+  not start failing as a result of this feature moving to GA.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+  To be documented as part of manual GA testing.
```

Comment:
We should have done that as part of Beta TBH (when we enable that by default).

Reply:
It may have been by the original authors of this KEP, I just don't have a record of it that I can include here. The feature moved to beta before this PRR section existed. I added a note to the graduation criteria for GA to follow up on this.

Reply:
I understand - just please ensure that this will be really tested and updated in this KEP before you promote feature-gate to GA in k/k code.